File: backend/services/OllamaService.js
const config = require('../config');
const Recipe = require('../models/Recipe');

class OllamaService {
  static async generateResponse(message, conversationHistory = [], userIngredients = []) {
    try {
      const systemPrompt = this.buildSystemPrompt(userIngredients);
      const conversationContext = this.buildConversationContext(
        systemPrompt, 
        conversationHistory, 
        message
      );
      
      const response = await fetch(`${config.ollama.url}/api/generate`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          model: config.ollama.model,
          prompt: conversationContext,
          stream: false,
          options: {
            temperature: config.ollama.temperature,
            num_predict: config.ollama.maxTokens
          }
        }),
        signal: AbortSignal.timeout(config.ollama.timeout)
      });

      if (!response.ok) {
        throw new Error(`Ollama API request failed: ${response.status}`);
      }

      const data = await response.json();
      return data.response || 'Sorry, I had trouble generating a response.';
      
    } catch (error) {
      console.error('Ollama service error:', error);
      
      if (error.name === 'TimeoutError' || error.message.includes('timeout')) {
        throw new Error('AI response timed out. Please try again with a shorter question.');
      }
      
      throw error;
    }
  }

  static buildSystemPrompt(userIngredients) {
    const ingredientsContext = userIngredients.length > 0 
      ? `\n\nUSER'S CURRENT INGREDIENTS: ${userIngredients.join(', ')}\nUse this information to provide personalized recipe suggestions, substitutions, and cooking advice.`
      : '\n\nThe user hasn\'t selected any ingredients yet, so focus on general cooking advice and encourage them to add ingredients.';

    const recipes = Recipe.getAll();
    const recipeList = recipes.map(recipe => 
      `- ${recipe.title} (${recipe.category}): ${recipe.ingredients.join(', ')}`
    ).join('\n');

    return `You are InstaDish, an expert cooking assistant and sustainability educator. You help users with:

COOKING EXPERTISE:
- Cooking techniques and methods
- Ingredient substitutions and alternatives  
- Food safety and storage tips
- Recipe modifications and improvements
- Troubleshooting cooking problems
- Kitchen tips and tricks

SUSTAINABILITY EDUCATION:
- Environmental impact of food choices
- Sustainable cooking practices
- Food waste reduction
- Carbon footprint of ingredients
- Local and seasonal eating
- Plant-based alternatives

RECIPE DATABASE (${recipes.length} recipes available):
${recipeList}
${ingredientsContext}

IMPORTANT INSTRUCTIONS:
1. Always consider the user's current ingredients when giving advice
2. Suggest specific recipes from our database that match their ingredients
3. Provide ingredient substitutions based on what they have available
4. Calculate and explain sustainability scores for their ingredient combinations
5. Be encouraging, educational, and explain the "why" behind suggestions
6. If they mention ingredients not in their list, suggest adding them for better matches
7. Provide specific cooking tips and techniques
8. Always mention sustainability impact when relevant

SUSTAINABILITY SCORING SYSTEM:
- Beef/Lamb: -25 points (very high carbon footprint: 27-21kg CO₂/kg)
- Cheese: -20 points (high carbon footprint: 21kg CO₂/kg)
- Pork: -15 points (medium-high carbon footprint: 12kg CO₂/kg)
- Chicken: -10 points (moderate carbon footprint: 7kg CO₂/kg)
- Eggs: -5 points (low carbon footprint: 5kg CO₂/kg)
- Rice/Grains: +5 points (low carbon footprint: 4kg CO₂/kg)
- Vegetables/Fruits: +15 points (very low carbon footprint: 1-2kg CO₂/kg)
- Legumes/Beans/Lentils: +20 points (excellent sustainability: 1kg CO₂/kg)

RESPONSE STYLE:
- Be conversational and friendly
- Use emojis appropriately (🍳🥗🌱♻️)
- Provide actionable advice
- Always explain the reasoning behind suggestions
- Keep responses informative but concise`;
  }

  static buildConversationContext(systemPrompt, conversationHistory, message) {
    let context = systemPrompt + "\n\nCONVERSATION:\n";
    
    if (conversationHistory && conversationHistory.length > 0) {
      conversationHistory.slice(-6).forEach(msg => {
        const role = msg.sender === 'user' ? 'User' : 'Assistant';
        context += `${role}: ${msg.text}\n`;
      });
    }
    
    context += `User: ${message}\nAssistant:`;
    
    return context;
  }

  static getSuggestedRecipes(userIngredients, limit = 3) {
    if (!userIngredients || userIngredients.length === 0) {
      return [];
    }

    const recipes = Recipe.getAll();
    
    const matchingRecipes = recipes.map(recipe => {
      const match = recipe.calculateMatch(userIngredients);
      return {
        ...recipe,
        match
      };
    })
    .filter(recipe => recipe.match.matches > 0)
    .sort((a, b) => b.match.percentage - a.match.percentage)
    .slice(0, limit);
    
    return matchingRecipes;
  }

  static async isOllamaAvailable() {
    try {
      const response = await fetch(`${config.ollama.url}/api/tags`, {
        method: 'GET',
        signal: AbortSignal.timeout(5000)
      });
      return response.ok;
    } catch (error) {
      console.error('Ollama availability check failed:', error);
      return false;
    }
  }

  static async getAvailableModels() {
    try {
      const response = await fetch(`${config.ollama.url}/api/tags`, {
        method: 'GET',
        signal: AbortSignal.timeout(5000)
      });
      
      if (!response.ok) {
        throw new Error('Failed to fetch models');
      }
      
      const data = await response.json();
      return data.models || [];
    } catch (error) {
      console.error('Error fetching Ollama models:', error);
      return [];
    }
  }
}

module.exports = OllamaService;

File: backend/services/ollama_service.py
import httpx
import json
import os
from typing import List, Dict, Any, Optional
from dotenv import load_dotenv
from pathlib import Path

# Load environment variables from root .env file
load_dotenv(dotenv_path=Path(__file__).parent.parent.parent / ".env")

class OllamaService:
    def __init__(self, base_url: str = None):
        self.base_url = base_url or os.getenv("OLLAMA_URL", "http://localhost:11434")
        self.model = os.getenv("OLLAMA_MODEL", "llama2:7b")
        
    async def generate_response(self, message: str, user_ingredients: List[str], relevant_recipes: List[Dict[str, Any]], selected_recipe: Dict[str, Any] = None) -> str:
        """Generate AI response using Ollama with RAG context"""
        try:
            # Build system prompt with recipe context
            system_prompt = self._build_system_prompt(user_ingredients, relevant_recipes, selected_recipe)
            
            # Prepare the full prompt
            full_prompt = f"{system_prompt}\n\nUser: {message}\n\nAssistant:"
            
            # Call Ollama API
            async with httpx.AsyncClient(timeout=60.0) as client:
                response = await client.post(
                    f"{self.base_url}/api/generate",
                    json={
                        "model": self.model,
                        "prompt": full_prompt,
                        "stream": False,
                        "options": {
                            "temperature": 0.7,
                            "top_p": 0.9,
                            "max_tokens": 500
                        }
                    }
                )
                
                if response.status_code == 200:
                    result = response.json()
                    return result.get("response", "I'm sorry, I couldn't generate a response.")
                else:
                    return f"I'm having trouble connecting to the AI service (Status: {response.status_code}). Please try again later."
                    
        except httpx.TimeoutException:
            return "I'm taking a bit longer to respond. Please try again with a shorter message."
        except httpx.ConnectError:
            return "I'm having trouble connecting to the AI service. Please make sure Ollama is running."
        except Exception as e:
            return f"I'm sorry, I encountered an error: {str(e)}"
    
    def _build_system_prompt(self, user_ingredients: List[str], relevant_recipes: List[Dict[str, Any]], selected_recipe: Dict[str, Any] = None) -> str:
        """Build system prompt with recipe context"""
        prompt = """You are InstaDish, a professional culinary assistant. 
                    Provide precise, accurate cooking advice and recipe recommendations.
                    Be concise and informative in your responses.
Current context:"""
        
        if user_ingredients:
            prompt += f"\nUser's ingredients: {', '.join(user_ingredients)}"
        
        if selected_recipe:
            prompt += f"\n\n🎯 SELECTED RECIPE: {selected_recipe['name']}"
            prompt += f"\nCategory: {selected_recipe['category']}"
            prompt += f"\nPrep time: {selected_recipe['prep_time']}"
            prompt += f"\nCook time: {selected_recipe['cook_time']}"
            prompt += f"\nDifficulty: {selected_recipe['difficulty']}"
            if selected_recipe.get('ingredients'):
                prompt += f"\nIngredients: {', '.join(selected_recipe['ingredients'][:5])}"
            if selected_recipe.get('sustainability'):
                sustain = selected_recipe['sustainability']
                prompt += f"\nSustainability: {sustain['level'].upper()} (Score: {sustain['score']}/3)"
            prompt += "\n\nFocus your advice on this selected recipe! Provide cooking tips, substitutions, and detailed guidance. Also, consider similar recipes in case the user asks."
        
        if relevant_recipes and not selected_recipe:
            prompt += "\nRelevant recipes:"
            for i, recipe in enumerate(relevant_recipes[:2], 1):
                prompt += f"\n{i}. {recipe['name']} - {recipe['category']}"
                if recipe.get('match'):
                    match = recipe['match']
                    prompt += f" ({match['percentage']}% match)"
        
        prompt += "\n\nBe helpful and respond to the question asked. Don't be too chatty. IMPORTANT: Always use line breaks (\\n) to separate different points, steps, or items. Never write everything in one long paragraph. Use proper formatting with line breaks for better readability."
        
        return prompt
    
    async def is_available(self) -> bool:
        """Check if Ollama service is available"""
        try:
            async with httpx.AsyncClient(timeout=5.0) as client:
                response = await client.get(f"{self.base_url}/api/tags")
                return response.status_code == 200
        except:
            return False
